File: containers/metadataparser/create-aggregate-qc-file.R
#!/usr/bin/env Rscript

## Tabulate alignment metrics from Alfred and CollectHsMetrics across multiple samples
## author  = Philip Jonsson
## contributor = Evan Biederstedt, evan.biederstedt@gmail.com, 2019 ; Anne Marie Noronha, noronhaa@mskcc.org, 2020

## email   = jonssonp@mskcc.org, evan.biederstedt@gmail.com, noronhaa@mskcc.org
## version = 0.2.2
## status  = Dev

suppressPackageStartupMessages({
    library(data.table)
    library(parallel)
    library(argparse)
})

#args = commandArgs(TRUE)

#if (is.null(args) | length(args)<1) {
#    message('Usage: create-aggregate-qc-file.R assay\n')
#    quit()
#} 
## default values now exist for each argument. error message is no longer necessary.

parser = ArgumentParser(description = 'Aggregate QC metrics across samples')
parser$add_argument('-n', '--num-cores', type = 'integer', required = TRUE, default = 1,
                    help = 'number of cores')
parser$add_argument('-a', '--assay-type', type = 'character', required = TRUE, default = 'wes',
                    help = 'Assay type')
args = parser$parse_args()

## Parse Alfred output ----------------------------------------------------------------------------------------------
read_alfred_qc = function(file) {
    input_file = fread(cmd = paste('zgrep ^ME', file, '| cut -f 2-')) 
    ## define column TotalReads
    input_file[, TotalReads := round(`#Mapped`/MappedFraction)]
    ##
    ## only select columns
    ## c('Sample', 'TotalReads','#DuplicateMarked','DuplicateFraction','#Mapped','MappedFraction','#Unmapped',
    ##     'UnmappedFraction','#SecondaryAlignments','SecondaryAlignmentFraction','#Pairs','#MappedProperPair','MappedProperFraction','MedianReadLength','MedianInsertSize','MedianCoverage')
    ## 
    ## For clarity, doing this in two steps
    ## subset columns
    input_file = input_file[, c('Sample', 'TotalReads','#DuplicateMarked','DuplicateFraction','#Mapped','MappedFraction',
          '#Unmapped','UnmappedFraction','#SecondaryAlignments','SecondaryAlignmentFraction','#Pairs',
          '#MappedProperPair','MappedProperFraction','MedianReadLength','MedianInsertSize','MedianCoverage')]

    setnames(input_file, 
        old = c('#DuplicateMarked', 'DuplicateFraction', '#Mapped', 'MappedFraction', '#Unmapped', 'UnmappedFraction', '#SecondaryAlignments', 'SecondaryAlignmentFraction', '#Pairs','#MappedProperPair', 'MappedProperFraction'), 
        new = c('NumberDuplicateMarked', 'FractionDuplicateMarked', 'NumberMapped', 'FractionMapped', 'NumberUnmapped', 'FractionUnmapped', 'NumberSecondaryAlignments', 'FractionSecondaryAlignments', 'TotalPairs', 'NumberMappedProperPairs', 'FractionMappedProperPairs'))
   
    return(input_file)
} 


## Read CollectHsMetrics output ------------------------------------------------------------------------------------
read_hs_metrics = function(file) {
    input_file = fread(cmd = paste('grep -A2 \"## METRICS\"', file, '| tail -n +2')) 
    ### define column Sample
    input_file[, Sample := gsub('.hs_metrics.txt', '', file)]
    ## For clarity, doing this in two steps
    ## subset columns
    input_file = input_file[, c("Sample", "ON_BAIT_BASES", "ON_TARGET_BASES", "MEAN_BAIT_COVERAGE", "MEAN_TARGET_COVERAGE", "MEDIAN_TARGET_COVERAGE", "MAX_TARGET_COVERAGE",
        "ZERO_CVG_TARGETS_PCT", "PCT_TARGET_BASES_10X", "PCT_TARGET_BASES_20X", "PCT_TARGET_BASES_50X", "PCT_TARGET_BASES_100X")]

    setnames(input_file, 
        old = c("Sample", "ON_BAIT_BASES", "ON_TARGET_BASES", "MEAN_BAIT_COVERAGE", "MEAN_TARGET_COVERAGE", "MEDIAN_TARGET_COVERAGE", "MAX_TARGET_COVERAGE", "ZERO_CVG_TARGETS_PCT", "PCT_TARGET_BASES_10X", "PCT_TARGET_BASES_20X", "PCT_TARGET_BASES_50X", "PCT_TARGET_BASES_100X"), 
        new = c("Sample", "BasesOnBait", "BasesOnTarget", "MeanBaitCoverage","MeanTargetCoverage","MedianTargetCoverage", "MaxTargetCoverage", "FractionTargetsZeroCoverage", "FractionTargets10X", "FractionTargets20X", "FractionTargets50X", "FractionTargets100X"))

    return(input_file)
}

read_wgs_metrics = function(file) {
    input_file = fread(cmd = paste('grep -A2 \"## METRICS\"', file, '| tail -n +2')) 
    input_file[, Sample := gsub('.wgs_metrics.txt', '', file)]
    input_file = input_file[,c("Sample","MEAN_COVERAGE","SD_COVERAGE","MEDIAN_COVERAGE","PCT_EXC_TOTAL","PCT_1X","PCT_10X","PCT_20X","PCT_50X","PCT_100X")]
    
    setnames(input_file, 
        old = c("Sample","MEAN_COVERAGE","SD_COVERAGE",               "MEDIAN_COVERAGE","PCT_EXC_TOTAL",        "PCT_1X","PCT_10X","PCT_20X","PCT_50X","PCT_100X"), 
        new = c("Sample","MeanCoverage", "StandardDeviationOfCoverage","MedianCoverage","FractionExcludedBases","FractionBases1X","FractionBases10X","FractionBases20X","FractionBases50X","FractionBases100X"))

    return(input_file)
}

## Execute in run directory ----------------------------------------------------------------------------------------

if (!interactive()) {
    
    # Run-time parameters
    num_cores = args$num_cores
    assay_type = args$assay_type
    
    # Parse output from Alfred 
    alfred_files = dir(getwd(), pattern = '*.alfred.tsv.gz$')
    outMatrix = mclapply(alfred_files, read_alfred_qc, mc.cores = num_cores)
    outMatrix = as.data.table(rbindlist(outMatrix, fill=TRUE)) ## same as do.call('rbind', ... )
    
    # For exomes, parse output from CollectHsMetrics
    if (assay_type == 'wes') {
        hs_metrics_files = dir(getwd(), pattern = '*.hs_metrics.txt$')
        hs_metrics_out = mclapply(hs_metrics_files, read_hs_metrics, mc.cores = num_cores) 
        hs_metrics_out = as.data.table(rbindlist(hs_metrics_out, fill=TRUE))   ## same as do.call('rbind', ... )
        setkey(outMatrix, Sample)
        setkey(hs_metrics_out, Sample)
        outMatrix = as.data.table(merge(outMatrix, hs_metrics_out, by = 'Sample'))
        outMatrix[, MedianCoverage:=NULL]
    } else if (assay_type == 'wgs'){
        wgs_metrics_files = dir(getwd(), pattern = '*.wgs_metrics.txt$')
        wgs_metrics_out = mclapply(wgs_metrics_files, read_wgs_metrics, mc.cores = num_cores) 
        wgs_metrics_out = as.data.table(rbindlist(wgs_metrics_out, fill=TRUE))   ## same as do.call('rbind', ... )
        setkey(outMatrix, Sample)
        setkey(hs_metrics_out, Sample)
        outMatrix = as.data.table(merge(outMatrix, wgs_metrics_out, by = 'Sample'))
        outMatrix[, MedianCoverage:=NULL]
    }
    
    fwrite(outMatrix, 'alignment_qc.txt', sep = '\t')

}

File: containers/metadataparser/create-aggregate-qc-file.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
 
"""
Custom aggregation of data from Alfred and GATK CollectHSMetrics.
Usage: filter-vcf.py assay
Output: 'input_filename.filter.vcf'
"""

__author__  = "Philip Jonsson"
__email__   = "jonssonp@mskcc.org"
__version__ = "0.1.0"
__status__  = "Dev"

import sys, glob, subprocess, multiprocessing
import pandas as pd
from io import StringIO
from joblib import Parallel, delayed

numCores = multiprocessing.cpu_count()
assayType = sys.argv[1]

def main():
    alfredFiles = glob.glob('*.alfred.tsv.gz')

    alfredOut = Parallel(n_jobs = numCores)(delayed(readAlfredQC)(fl) for fl in alfredFiles)
    outMatrix = pd.concat(alfredOut)

    if assayType == "wes":
        hsMetricsFiles = glob.glob('*.hs_metrics.txt')
        hsMetricsOut = Parallel(n_jobs = numCores)(delayed(readHsMetrics)(fl) for fl in hsMetricsFiles)
    
        hsMetricsOutMatrix = pd.concat(hsMetricsOut)
        outMatrix = pd.merge(outMatrix, hsMetricsOutMatrix, on = "Sample")

        outMatrix = outMatrix.drop(columns = "MedianCoverage")

    outMatrix.to_csv("alignment_qc.tsv", sep = "\t", index = False)

def readAlfredQC(file):
    readCmd = "zgrep ^ME %s | cut -f 2-" % file
    proc = subprocess.Popen(readCmd, shell = True, stdout = subprocess.PIPE)
    a = StringIO(proc.communicate()[0].decode('utf-8'))
    b = pd.read_csv(a, sep = "\t")
    b["TotalReads"] = round(b["#Mapped"]/b["MappedFraction"])

    c = b.rename(columns = {
        "#DuplicateMarked": "NumberDuplicateMarked",
        "DuplicateFraction": "FractionDuplicateMarked",
        "#Mapped": "NumberMapped",
        "MappedFraction": "FractionMapped",
        "#Unmapped": "NumberUnmapped",
        "UnmappedFraction": "FractionUnmapped",
        "#SecondaryAlignments": "NumberSecondaryAlignments",
        "SecondaryAlignmentFraction": "FractionSecondaryAlignments",
        "#Pairs": "TotalPairs",
        "#MappedProperPair": "NumberMappedProperPairs",
        "MappedProperFraction": "FractionMappedProperPairs"
    })
        
    return c[["Sample", "TotalReads", "NumberDuplicateMarked", "FractionDuplicateMarked", "NumberMapped", "FractionMapped", "NumberUnmapped", "FractionUnmapped", "NumberSecondaryAlignments", "FractionSecondaryAlignments", "TotalPairs", "NumberMappedProperPairs", "FractionMappedProperPairs", "MedianReadLength", "MedianInsertSize", "MedianCoverage"]]

def readHsMetrics(file):
    readCmd = "grep -A2 \"## METRICS\" %s | tail -n +2" % file
    proc = subprocess.Popen(readCmd, shell = True, stdout = subprocess.PIPE)
    a = StringIO(proc.communicate()[0].decode('utf-8'))
    b = pd.read_csv(a, sep = "\t")

    b['Sample'] = file.split(".hs_metrics.txt")[0]

    c = b.rename(columns = {
        "ON_BAIT_BASES": "BasesOnBait",
        "ON_TARGET_BASES": "BasesOnTarget",
        "MEAN_BAIT_COVERAGE": "MeanBaitCoverage",
        "MEAN_TARGET_COVERAGE": "MeanTargetCoverage",
        "MEDIAN_TARGET_COVERAGE": "MedianTargetCoverage",
        "MAX_TARGET_COVERAGE": "MaxTargetCoverage",
        "ZERO_CVG_TARGETS_PCT": "FractionTargetsZeroCoverage",
        "PCT_TARGET_BASES_10X": "FractionTargets10X",
        "PCT_TARGET_BASES_20X": "FractionTargets20X",
        "PCT_TARGET_BASES_50X": "FractionTargets50X",
        "PCT_TARGET_BASES_100X": "FractionTargets100X"
    })

    return c[["Sample", "BasesOnBait", "BasesOnTarget", "MeanBaitCoverage", "MeanTargetCoverage", "MedianTargetCoverage", "MaxTargetCoverage", "FractionTargetsZeroCoverage", "FractionTargets10X", "FractionTargets20X", "FractionTargets50X", "FractionTargets100X"]]

if __name__ == "__main__":
    main()
